src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1078,27 +1078,46 @@
                            fact-binding (conj fact-binding))
 
             ;; Find children that all parent nodes have.
-            common-children-ids (into
-                                 ;; Since we require that id-to-condition-nodes have an equal value to "node" under the ID
-                                 ;; for this to be used, the order of this set probably doesn't matter in most cases.  However, in any possible edge
-                                 ;; cases where there are equal nodes under different IDs, sorting the set will add determinism.
-                                 ;; Having different nodes under the same
-                                 ;; ID would be a bug, but having an equivalent node under multiple IDs wouldn't necessarily be one.
-                                 (sorted-set)
-                                 (apply s/union (-> (:forward-edges beta-graph)
-                                                    (select-keys parent-ids)
-                                                    (vals))))
-
+            forward-edges (if (= 1 (count parent-ids))
+                            ;; If there is only one parent then there is no need to reconstruct
+                            ;; the set of forward edges. This is an intentional performance optimization for large
+                            ;; beta graphs that have many nodes branching from a common node, typically the root node.
+                            (-> (:forward-edges beta-graph)
+                                (get (first parent-ids)))
+                            (->> (select-keys parent-ids (:forward-edges beta-graph))
+                                 vals
+                                 ;; Order doesn't matter here as we will effectively sort it using update-node->id later,
+                                 ;; thus adding determinism.
+                                 (into #{} cat)))
```

Comment:
This is probably marginally faster than your previous version:

````
(into [] (comp cat (distinct)))
````
since that one conj'ed on every distinct element to a persistent set (thus creating lots of sets that will have to be GC'ed) as well as building a vector via transients, while this new version will just build a set via transients.  See [this optimization in into](https://github.com/clojure/clojure/blob/clojure-1.9.0/src/clj/clojure/core.clj#L6820).

Nevertheless, since there was a non-comment change since the last iteration, I'd like to make sure we've benchmarked this final version against the test we were using for the previous version - did you do so @EthanEChristian ?  The main benefit here might not show up in a benchmark easily, since it would be to decrease the number of objects created that might not be GC'ed until later, but I'd like to make sure we didn't inadvertently cause a regression.

Reply:
For the most part in the real world Sessions that i have been benchmarking this against, this change has decreased the time spent in add-conjunction slightly.

I took some time to profile the generative rules mentioned [here](https://github.com/cerner/clara-rules/pull/378#issuecomment-359600045), this time with 20000 productions.

The file below contains VM snapshots gathered using the current state of clara, versus the current state of this branch.
[ISSUE_377_NON_IP.zip](https://github.com/cerner/clara-rules/files/1755287/ISSUE_377_NON_IP.zip)